### PR TITLE
Add RSS feed

### PR DESCRIPTION
**Pull request by tttppp**

### CODE REVIEW

**Comment by tttppp** [comment]
This is pretty basic at the moment, but I wanted to check the concept first.  Also I wanted to find out how to use memcache caching.

**Comment by tttppp** [comment]
I had a look at Memcache in the App Engine, but it wasn't the sort of cache I was expecting.  I've added application-level caching in the form of Etag, Last-Modified and Cache-Control.  RSS readers should be honouring one or more of these.  (My implementation for "Last-Modified" is a bit brutal).

Is Memcache the GCP service to add a cache for a URL?  I couldn't see how to use it for this.

**Comment by zond** [comment]
Nice caching layer!

The App Engine memcache is primitive - you have to build your own semantics on top.

E.g. every time you load a particular feed, you first try to load it from memcache. If it succeeds you serve it. If not you load from db and then put it in memcache with an e.g. five minute ttl.

The memcache key you use is then created so that it contains all relevant parameters of your db query, so that each query gets its own cache.

**Comment by zond** [comment]
Anyway, I'm fine with this as it is now, commit when you feel ready :)

**Comment by tttppp** [comment]
I might have a go at using the memcache then. I realised that if a feed is being queried for a finished game, then the Last-Modified date will always be more than 24 hours ago, and so the db will always get hit. RSS readers should honour the Cache-Control and only ask once a day, but from what I read there are a some that just use Last-Modified.

It might make sense to have special behaviour for feeds that just contain finished games.

**Comment by tttppp** [comment]
This should be a good start though.

**Comment by zond** [comment]
Ah yeah, set Cache-Control to a year or something for finished games? :)

Or maybe a week, in case we want to change some details.

**Comment by tttppp** [comment]
Now with Cache-Control set to a year for finished games (great idea, thanks!), and a couple of dates stored in memcache per feed.

The modified and checked dates in memcache allow much more sensible use of If-Modified-Since.  Due to the maximum size limit on memcache objects it didn't make sense to cache the whole feed, but as I mentioned earlier, RSS readers should be fairly good about not DoS-ing feed servers.

**Comment by zond** [comment]
Nice!

Note that a common trick is to zip the bytes before you put them in memcache, that way you might possibly fit any game completely in memcache.

You don't HAVE to, but it might work :)

File: game/rss.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+			}
+		}
+	}
+
+	// Process the request.
+	uq := r.Req().URL.Query()
+
+	limit, err := strconv.ParseInt(uq.Get("limit"), 10, 64)
+	if err != nil || limit > maxLimit {
+		limit = maxLimit
```
Comment (tttppp) [inline game/rss.go]: This doesn't seem to be used.

File: game/rss.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		if phaseTypeFilter := uq.Get("phaseType"); phaseTypeFilter != "" {
+			q = q.Filter("Type=", phaseTypeFilter)
+		}
+
+		if _, err := q.GetAll(ctx, &phases); err != nil {
+			return err
+		}
+		for _, phase := range phases {
+			title := fmt.Sprintf("%s %d %s %s (%s)", game.Desc, phase.Year, phase.Season, phase.Type, game.Variant)
+			description := "Map should go here"
```
Comment (tttppp) [inline game/rss.go]: Perhaps include the SC count here as a placeholder?

**Comment by tttppp** [comment]
Right - I think this is done. Merging and seeing what happens...

**Comment by tttppp** [comment]
I'm not understanding how the index.yaml file is meant to work. I added some indexes to get it working locally, but I'm now getting complaints on the live server as if I didn't add any indexes:
https://github.com/zond/diplicity/pull/112/commits/2f3acdd64f04eb58b73bf343f3e7cf76435c4229

https://diplicity-engine.appspot.com/Rss?api-level=5&variant=Classical

Do I have to populate the new indexes somehow?

**Comment by tttppp** [comment]
I'd expect to see an index with kind Phase here: https://console.cloud.google.com/datastore/indexes?project=diplicity-engine

**Comment by zond** [comment]
I think there's a separate command to update index.yaml, cron.yaml etc. Check the help for the deploy command!

**Comment by tttppp** [comment]
Great - thanks!  The command was (logically) `gcloud app deploy index.yaml`. Just need to wait for the index to be created now.

**Comment by zond** [comment]
:)

**Pull request merged**